File: projects/BlogApi/python/blog_api/posts.py
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, Field

from .auth import CurrentUser, get_current_user

router = APIRouter(tags=["posts"])


class PostBase(BaseModel):
    title: str = Field(min_length=3)
    body: str = Field(min_length=10)


class PostCreate(PostBase):
    pass


class Post(PostBase):
    id: str
    author_id: str


_POSTS: dict[str, Post] = {}
_SEQ = 0


def _next_id() -> str:
    global _SEQ
    _SEQ += 1
    return f"post_{_SEQ}"


@router.get("/posts", response_model=dict)
def list_posts(limit: int = 20, cursor: Optional[str] = None):
    posts = list(_POSTS.values())
    posts = posts[: min(limit, 50)]
    return {
        "data": posts,
        "pagination": {
            "next_cursor": None,
            "previous_cursor": None,
            "limit": limit,
        },
    }


@router.post(
    "/posts",
    response_model=Post,
    status_code=status.HTTP_201_CREATED,
)
def create_post(
    payload: PostCreate,
    current_user: CurrentUser = Depends(get_current_user),
):
    post = Post(
        id=_next_id(),
        title=payload.title,
        body=payload.body,
        author_id=current_user.user_id,
    )
    _POSTS[post.id] = post
    return post

File: projects/BlogApi/typescript/src/routes/posts.ts
import express from 'express';
import { PostService } from '../services/PostService';

export function createPostsRouter(postService: PostService): express.Router {
  const router = express.Router();

  router.get('/posts', (req, res, next) => {
    try {
      const limit = Math.min(50, Number(req.query.limit ?? 20));
      const posts = postService.listPosts(limit);
      res.status(200).json({
        data: posts,
        pagination: {
          next_cursor: null,
          previous_cursor: null,
          limit,
        },
      });
    } catch (err) {
      next(err);
    }
  });

  router.post('/posts', (req, res, next) => {
    try {
      const { title, body, author_id } = req.body ?? {};

      const errors: Record<string, string[]> = {};
      if (!title) errors.title = ['This field is required.'];
      if (!body) errors.body = ['This field is required.'];
      if (!author_id) errors.author_id = ['This field is required.'];

      if (Object.keys(errors).length > 0) {
        return res.status(422).json({
          error: {
            code: 'VALIDATION_ERROR',
            message: 'Validation failed.',
            details: errors,
          },
        });
      }

      const post = postService.createPost({
        title,
        body,
        authorId: author_id,
      });

      res.status(201).json(post);
    } catch (err) {
      next(err);
    }
  });

  return router;
}
